modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -107,27 +119,52 @@ public Netty4HttpPipeliningHandler(
 
     @Override
     public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
-        activityTracker.startActivity();
-        try {
-            assert msg instanceof FullHttpRequest : "Should have fully aggregated message already but saw [" + msg + "]";
-            final FullHttpRequest fullHttpRequest = (FullHttpRequest) msg;
-            final Netty4HttpRequest netty4HttpRequest;
-            if (fullHttpRequest.decoderResult().isFailure()) {
-                final Throwable cause = fullHttpRequest.decoderResult().cause();
-                final Exception nonError;
-                if (cause instanceof Error) {
-                    ExceptionsHelper.maybeDieOnAnotherThread(cause);
-                    nonError = new Exception(cause);
+        if (msg instanceof HttpRequest request) {
+            try {
+                activityTracker.startActivity();
+                final Netty4HttpRequest netty4HttpRequest;
+                if (request.decoderResult().isFailure()) {
+                    final Throwable cause = request.decoderResult().cause();
+                    final Exception nonError;
+                    if (cause instanceof Error) {
+                        ExceptionsHelper.maybeDieOnAnotherThread(cause);
+                        nonError = new Exception(cause);
+                    } else {
+                        nonError = (Exception) cause;
+                    }
+                    netty4HttpRequest = new Netty4HttpRequest(readSequence++, (FullHttpRequest) request, nonError);
                 } else {
-                    nonError = (Exception) cause;
+                    if (request instanceof FullHttpRequest fullHttpRequest) {
+                        currentRequestStream = null;
+                        netty4HttpRequest = new Netty4HttpRequest(readSequence++, fullHttpRequest);
+                    } else {
+                        var contentStream = new Netty4HttpRequestBodyStream(ctx.channel());
+                        currentRequestStream = contentStream;
```

Comment:
Will add assertion. It should never happen, all HTTP parts should be properly ordered and framed.

Reply:
> It can be non-null if previous request was stream too

We [set](https://github.com/elastic/elasticsearch/pull/111438/files#diff-21f42e3b52d200668ccaefd2de2d556720e1ad5cc86b9259887805b3dc36c117R165-R166) `currentRequestStream` back to `null` on receiving `LastHttpContent`. Or do you mean that may not necessary happen for every streaming request?

Reply:
Right, sorry I lost track of my own changes :) In normal circumstances should be null. If request is not properly terminated (no last content) and there is new request then `currentRequestStream` might be not null, but it will end up with decoding failure and connection shutdown. I will add test.